adi/cn0554.py

```diff
@@ -0,0 +1,148 @@
+# PARTICULAR PURPOSE ARE DISCLAIMED.
+#
+# IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, INTELLECTUAL PROPERTY
+# RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+# BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import adi
+
+
+class cn0554:
+    """ CN0554 Mixed Signal Raspberry Pi Hat """
+
+    _in_range_available = ["+/-13.75", "+27.5", "+2.5"]
+    _out_reference_available = [4.096, 2.5]
+    _out_channels = 16
+    _in_channels = 8
+
+    def __init__(self, uri="ip:analog.local"):
+        self.scale_res = 11  # Based on resistors used in input
+        self.adc = adi.ad7124(uri=uri)
+        self.adc._rx_unbuffered_data = False
+        self.adc._rx_output_type = "SI"
+        self.data_points = 1024
+        self.adc.sample_rate = 19200
+        self.adc_in_channels = []
+
+        for in_ch in self.adc.channel:
+            in_name = "voltage"
+            for char in in_ch.name:
+                if char.isnumeric():
+                    in_name += char
+
+            setattr(
+                self, in_name + "_in_range", self._in_range_available[0]
+            )  # default range of HW
+
+            self.adc_in_channels.append(in_name)
+
+        self.dac = adi.ltc2688(uri=uri)
+        self.dac_out_channels = self.dac.channel_names
+        self.out_reference = self._out_reference_available[0]
+
+    @property
+    def out_type(self):
+        """ Get value of CN0554's rx output type (raw or SI) """
+        return self.adc._rx_output_type
+
+    @out_type.setter
+    def out_type(self, val):
+        """ Set value of CN0554's rx output type (raw or SI) """
+        self.adc._rx_output_type = val
+
+    @property
+    def enabled_in_channels(self):
+        """ Get list of enabled input adc channels """
+        return self.adc.rx_enabled_channels
+
+    @enabled_in_channels.setter
+    def enabled_in_channels(self, val):
+        """ Set list of enabled input adc channels """
+        self.adc.rx_enabled_channels = val
+
+    @property
+    def out_channels(self):
+        """ Get number of DAC output channels """
+        return self._out_channels
+
+    @property
+    def in_channels(self):
+        """ Get number of ADC input channels """
+        return self._in_channels
+
+    @property
+    def out_reference(self):
+        """ Get voltage reference used for DAC channels """
+        return self.out_reference
+
+    @out_reference.setter
+    def out_reference(self, value):
+        """ Set voltage reference used for DAC channels """
+        if value in self._out_reference_available:
+            self.dac.vref = value
```

Comment:
Changed the property function for out_reference to return directly the self.dac.vref value to avoid recursion. Rewrote the setter method to check if value is valid and raise an exception if not.